# How to use autodif:

**Install autodiff package:**

----
*Option 1:*

Our package is available on PyPI. The preferred way for installing our package via PyPI is as follows:

First, the user can (optionally) create a virtual environment before installing our package dependencies. A virtual enviroment would allow the user to compartmentalize the the dependencies for our package from the dependencies of other projects they might be working on.

For users unfamiliar with virtual environments, the tool virtualenv can be easily installed via:

    sudo easy_install virtualenv

The user can then activate a virtual environment via:

    virtualenv env
    source env/bin/activate
    
Note: the user can later deactivate the virtual enviroment via:

    deactivate
    
Then, the user can simply install our package from PyPI via:

    pip install autodiff-ADdictedtoCS

----
*Option 2:*

Alternatively, the autodiff package can be downloaded directly from our organization's github repository at: https://github.com/ADdictedtoCS/cs207-FinalProject.git

The user can then install the requirements via:

    pip install -r requirements.txt

The user can then import our package as follows...

In [2]:
#  Import statements
import autodiff
import autodiff.function as F
import autodiff.optim as optim
from autodiff.variable import Variable

See below for an example of how the user can interact with our package:

1. Create a "variable" instantiated with a vector of initial values
    - For a multivariate function of 3 inputs for example, "variable" takes in the initial value
      of each input
2. Define a function 
    - Functions take variables as inputs and return new variables with updated values and gradients
3. Call the function and get the value and gradient of the resulting variable 
    - The user can simply print the variable or alternatively call .val or .grad on the variable

*Remarks*: 
- The user can benefit from a natural and simple way to interact with the Variables and Functions.
-  autodiff.function also supports a range of different elementary functions. See "Implementation Details" section below for further details.


### Example 1: Single input, single output

In [5]:
# Example R^1 -> R^1

# Define a variable with an initial value
x = Variable(0.)
print("Input x:")
print(x)

# Define a function
def my_func(x):
    return F.sin(F.exp(x))

# Variable z is the result of calling function on x
z = my_func(x)

# Get value and gradient of z
print("\nOutput z:")
print(z)

# Alternatively:
print('The value is: {}'.format(z.val))
print('The gradient is: {}'.format(z.grad))

Input x:
Value:
0.0
Gradient:
1.0

Output z:
Value:
0.8414709848078965
Gradient:
0.5403023058681398
The value is: 0.8414709848078965
The gradient is: 0.5403023058681398


Above, we initialize a variable x with a value of 0

We then define the function:  
     $f(x) = sin(exp(x))$

Lastly, we get the value and the gradient of the function at the initial point:  
Value: $sin(exp(0)) = 0.841$  
Gradient: $exp(0)*cos(exp(0)) = 0.540$  

### Example 2: Multi input, single output

In [7]:
# Example R^3 -> R^1

# Define a variable with a vector of initial values
X = Variable([0,1,2])
print("Input X")
print(X)

# Define a function
# Unroll allows us to define an expression using the individual input variables
def my_func(X):
    x, y, z = X.unroll()
    return F.cos(x) + y * z

# Variable out is the result of calling function on X
out = my_func(X)

# Print value and gradient of the output
print('\nOutput:')
print('The value is: {}'.format(out.val))
print('The gradient is: {}'.format(out.grad))

Input X
Value:
[[0.]
 [1.]
 [2.]]
Gradient:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Output:
The value is: 3.0
The gradient is: [[0. 2. 1.]]


Above, we initialize "Variable" X with a vector of initial values [0, 1, 2]

X is thus a vector of 3 input variables, which can be obtained via "unroll".
We can call these input variables x, y, z

We then define a function in terms of the input variables  
     $f(x,y,z) = cos(x) + y*z$

Lastly, we get the value and the gradient of the function at the initial point:  
Value:  
$cos(0) + 1*2 = 3$  
Gradient:  
$ \frac{\partial{f}}{\partial{x}} = -sin(x) = 0$  
$ \frac{\partial{f}}{\partial{y}} = z = 2 $  
$ \frac{\partial{f}}{\partial{z}} = y = 1 $  

### Case 3: Multi input, multi output

In [11]:
# Example R^3 -> R^3

# Define a variable with a vector of initial values
X = Variable([0,1,2])
print("Input X")
print(X)

# Define a function
# Unroll allows us to define an expression using the individual input variables
# Concat is used to build the output vector of the function
def my_func3(X):
    x, y, z = X.unroll()
    o1 = F.sin(F.exp(x))
    o2 = y + F.cosh(z)
    o3 = F.Log(10)(z)
    out_X = F.concat([o1, o2, o3])
    return out_X

# Variable out is the result of calling function on X
out = my_func3(X)

# Print value and gradient of the output
print('\nOutput:')
print('The value is: \n{}'.format(out.val))
print('The gradient is: \n{}'.format(out.grad))

Input X
Value:
[[0.]
 [1.]
 [2.]]
Gradient:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Output:
The value is: 
[[0.84147098]
 [4.76219569]
 [0.30103   ]]
The gradient is: 
[[0.54030231 0.         0.        ]
 [0.         1.         3.62686041]
 [0.         0.         0.21714724]]
